### PR TITLE
Several small commits...

### PR DESCRIPTION
**Pull request by samfoo**
- Added some slightly better output for bt-go on errors
- Added a couple tests for diamond dependencies (they were already working, but tests make sense)
- Added some (brief) comments to the pipeline class.

### CODE REVIEW

File: lib/bt/pipeline.rb [hunk 1]
```diff
@@ -4,13 +4,18 @@ require 'set'
...
     def stages
-      stage_definition = self[:stage_definition].dup
+      unknown_stages = self[:stage_definition].dup
       Set.new.tap do |known_stages|
-        while !stage_definition.empty?
-          name, definition = next_satisfied! stage_definition, known_stages
+        while !unknown_stages.empty?
+          name, definition = next_satisfied! unknown_stages, known_stages
+
+          # Find all of the needs for this stage. They're already guaranteed to
```
Comment (quad) [inline lib/bt/pipeline.rb:12]: I'm glad you wrote this because hell if I understand what any of this does anymore. :-D

File: lib/bt/pipeline.rb [hunk 2]
```diff
@@ -4,13 +4,18 @@ require 'set'
 module BT
   class Pipeline < Struct.new(:commit, :stage_definition)
     def stages
-      stage_definition = self[:stage_definition].dup
+      unknown_stages = self[:stage_definition].dup
```
Comment (quad) [inline lib/bt/pipeline.rb:7]: Yeah, that's a better name for the variable.

**Comment by quad** [comment]
Why do we drop the .lock file when we have a Gemspec? Don't the versions of dependencies in development still matter?

File: bin/bt-go [hunk 3]
```diff
@@ -31,8 +31,15 @@ pipeline = BT::Pipeline.new commit, stage_definition
...
   stage = pipeline.stages.detect { |s| s.name == opts[:stage] } or raise "No stage #{opts[:stage].inspect} for commit #{commit.sha}"
-  raise "#{stage} is not ready" unless stage.ready?
-  stage.build
+
+  if stage.done?
+    puts "#{stage} is already done"
+  elsif stage.ready? 
+    stage.build
+  else
+    raise "#{stage} is not ready, it's blocked by #{stage.blockers.map(&:name).join(', ')}"
```
Comment (quad) [inline bin/bt-go:40]: OK, yes, obviously that is superior error output. God.

File: bin/bt-go [hunk 4]
```diff
@@ -31,8 +31,15 @@ pipeline = BT::Pipeline.new commit, stage_definition
 
 if opts[:stage]
   stage = pipeline.stages.detect { |s| s.name == opts[:stage] } or raise "No stage #{opts[:stage].inspect} for commit #{commit.sha}"
-  raise "#{stage} is not ready" unless stage.ready?
-  stage.build
+
+  if stage.done?
+    puts "#{stage} is already done"
```
Comment (quad) [inline bin/bt-go:36]: But, how do I rebuild an already done stage? Common use case: an environmental factor causes failure and I need to re-try.

File: lib/bt/pipeline.rb [hunk 5]
```diff
@@ -33,11 +38,13 @@ module BT
 
     private
 
-    def next_satisfied! stage_definition, known_stages
-      stage = stage_definition.detect do |name, definition|
+    def next_satisfied! unkown_stages, known_stages
```
Comment (quad) [inline lib/bt/pipeline.rb]: Typo on `unkown_stages`

File: lib/bt/stage.rb [hunk 6]
```diff
@@ -1,4 +1,4 @@
-module BT
+ module BT
```
Comment (quad) [inline lib/bt/stage.rb]: Accidental whitespace?

File: lib/bt/stage.rb [hunk 7]
```diff
@@ -33,11 +33,11 @@ module BT
     end
 
     def ok?
-      (r = result) && r.message.start_with?('PASS')
+      result && result.message.start_with?('PASS')
```
Comment (quad) [inline lib/bt/stage.rb]: `result.andand.message.start_with? 'PASS'` ?

We really need to decide if we're going to use `andand` or not.

File: spec/bt/pipeline_spec.rb [hunk 8]
```diff
@@ -77,6 +77,69 @@ first:
     end
   end 
 
+  context "with a diamond dependency" do
```
Comment (quad) [inline spec/bt/pipeline_spec.rb:80]: Word up. Bomb test.

File: spec/bt/pipeline_spec.rb [hunk 9]
```diff
@@ -113,9 +176,6 @@ first:
 
     describe 'first stage' do
       subject { pipeline.stages.detect {|s| s.name == 'first' } }
-
-      it { should have(1).needs }
```
Comment (quad) [inline spec/bt/pipeline_spec.rb:117]: Why is this removed?

File: lib/bt/stage.rb (hunk 6 above)
Comment (samfoo) [inline lib/bt/stage.rb]: Yeah, that's accidental.

File: bin/bt-go (hunk 4 above)
Comment (samfoo) [inline bin/bt-go:36]: It wasn't doing this anyway, since #ready? checks "&& !done?". I assumed this was a feature not a bug. Am I misunderstanding #ready?

File: spec/bt/pipeline_spec.rb (hunk 9 above)
Comment (samfoo) [inline spec/bt/pipeline_spec.rb:117]: The line underneath checks `its(:needs) { should == [third] }`. The "having one need" is implicit and redundant.

File: lib/bt/pipeline.rb (hunk 5 above)
Comment (samfoo) [inline lib/bt/pipeline.rb]: The perils of qwerty...

File: bin/bt-go (hunk 4 above)
Comment (quad) [inline bin/bt-go:36]: No, you're not. I'm mis-reading. (What a great maintainer I am!)

Right, so, at the moment it'll print off "is not ready" even for a stage that's already built.

I wonder how we'd been handling that up until this point. Deleting the blocking commit and re-trying I imagine. Lame. Lame. Lame.

This leaves me thinking it might be time to document some fundamental use cases. That was going to be the idea of the cookbooks, but those won't be comprehensive for issue resolution and other such fun.

OK, I'm down with this change. It's clearly superior.
